catkin_virtualenv/cmake/catkin_generate_virtualenv.cmake

```diff
@@ -17,42 +17,55 @@
 # You should have received a copy of the GNU General Public License
 # along with this program. If not, see <http://www.gnu.org/licenses/>.
 function(catkin_generate_virtualenv)
-  set(oneValueArgs PYTHON_VERSION PYTHON_VERSION_MAJOR USE_SYSTEM_PACKAGES ISOLATE_REQUIREMENTS)
+  set(oneValueArgs PYTHON_VERSION PYTHON_INTERPRETER USE_SYSTEM_PACKAGES ISOLATE_REQUIREMENTS LOCK_FILE)
   set(multiValueArgs EXTRA_PIP_ARGS)
   cmake_parse_arguments(ARG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
 
-  # Check if this package already has a virtualenv target before creating one
-  if(TARGET ${PROJECT_NAME}_generate_virtualenv)
-    message(WARNING "catkin_generate_virtualenv was called twice")
-    return()
+  ### Handle argument defaults and deprecations
+
+  if(DEFINED ARG_PYTHON_VERSION)
+    message(WARNING "PYTHON_VERSION has been deprecated, set 'PYTHON_INTERPRETER python${ARG_PYTHON_VERSION}' instead")
+    set(ARG_PYTHON_INTERPRETER "python${ARG_PYTHON_VERSION}")
   endif()
 
-  # Backwards compatibility for PYTHON_VERSION_MAJOR, overriding PYTHON_VERSION
-  if(DEFINED ARG_PYTHON_VERSION_MAJOR)
-    set(ARG_PYTHON_VERSION ${ARG_PYTHON_VERSION_MAJOR})
-    message(WARNING "PYTHON_VERSION_MAJOR has been deprecated, please set PYTHON_VERSION instead")
+  if(NOT DEFINED ARG_PYTHON_INTERPRETER)
+    set(ARG_PYTHON_INTERPRETER "python2")
   endif()
 
-  if(NOT DEFINED ARG_PYTHON_VERSION)
-    set(ARG_PYTHON_VERSION 2)
+  if(NOT DEFINED ARG_USE_SYSTEM_PACKAGES OR ARG_USE_SYSTEM_PACKAGES)
+    message(STATUS "Using system site packages")
+    set(venv_args "${venv_args} --use-system-packages")
   endif()
 
-  if(NOT DEFINED ARG_USE_SYSTEM_PACKAGES)
-    set(ARG_USE_SYSTEM_PACKAGES TRUE)
+  if(ARG_ISOLATE_REQUIREMENTS)
+    message(STATUS "Only using requirements from this catkin package")
+    set(freeze_args "${freeze_args} --no-deps")
   endif()
 
-  if(NOT DEFINED ARG_ISOLATE_REQUIREMENTS)
-    set(ARG_ISOLATE_REQUIREMENTS FALSE)
+  if(NOT DEFINED ARG_LOCK_FILE)
+    set(ARG_LOCK_FILE "${CMAKE_BINARY_DIR}/requirements.txt")
+    message(WARNING "Please define a LOCK_FILE relative to your sources, and commit together to prevent dependency drift.")
+  else()
+    set(ARG_LOCK_FILE "${CMAKE_SOURCE_DIR}/${ARG_LOCK_FILE}")
   endif()
 
   if (NOT DEFINED ARG_EXTRA_PIP_ARGS)
-    set(ARG_EXTRA_PIP_ARGS "-qq")
+    set(ARG_EXTRA_PIP_ARGS "-qq" "--retries 10" "--timeout 30")
```

Comment:
@garyservin, why did we need to build in `retry` when pip had retries all along...